ext/src/zpages/tracez_processor.cc

```diff
@@ -0,0 +1,37 @@
+#include "opentelemetry/ext/zpages/tracez_processor.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace ext {
+namespace zpages {
+
+  void TracezSpanProcessor::OnStart(opentelemetry::sdk::trace::Recordable &span) noexcept {
+    if (shutdown_signal_received_) return;
+    spans_.running.insert(static_cast<opentelemetry::sdk::trace::SpanData*>(&span));
```

Comment:
Is this thread safe?

Reply:
As a start, we added copy-on-write functionality in the GetSpanSnapshot() function where possible to minimize potential contention issues. This is just an initial iteration, and we can add more safety measures as desired.

On a related note, we're also currently working on some tests (which will be added [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/f1f54f0ad476b223051948cb444e6703f8e61fc3/ext/test/zpages/tracez_processor_test.cc#L468)) to check for thread safety issues including the race conditions; we'll be sure to mitigate failures accordingly based on those results. We weren't sure if this needed to be complete before sending an initial PR; is that something we want before merging?

Reply:
I don't know how copy-on-write is related here.
`running` is an STL container which doesn't have thread safety guarantee. If `running.insert` is called concurrently, the behavior is undefined.

Reply:
My understanding is that copy-on-write would reduce contention, so at least the used running spans copy won't be changed mid-call? I agree it would be a good idea to check thread safety, especially with that and how the completed span vector gets cleared when GetSnapshot() is called

Reply:
Where is `GetSpanSnapshot` called?

Reply:
`GetSpanSnapshot()` is called by the TraceZ data aggregator, which is a WIP. You can check it out here: https://github.com/kmanghat/opentelemetry-cpp/pull/4